scipy/optimize/_group_columns.pyx

```diff
@@ -0,0 +1,90 @@
+"""
+Cython implementation of columns grouping for finite difference Jacobian
+estimation. Used by ._numdiff.group_columns.
+"""
+
+import numpy as np
+
+cimport numpy as np
+from cpython cimport bool
+
+
+def group_dense(int m, int n, int [:, :] A):
+    cdef int [:, :] B = A.T
+
+    groups = -np.ones(n, dtype=np.int32)
+    cdef int [:] groups_v = groups
+    cdef int current_group = 0
+
+    cdef int i, j, k
+    cdef bool intersect
+
+    union = np.empty(m, dtype=np.int32)
+    cdef int [:] union_v = union
+
+    for i in range(n):
+        if groups_v[i] >= 0:
+            continue
+
+        groups_v[i] = current_group
+        non_grouped, = np.where(groups < 0)
+
+        union_v[:] = B[i]
+        if non_grouped.size == 0:
+            break
+
+        for j in non_grouped:
+            intersect = False
+            for k in range(m):
+                if union_v[k] > 0 and B[j, k] > 0:
+                    intersect = True
+                    break
+
+            if not intersect:
+                union += B[j]
+                groups_v[j] = current_group
+
+        current_group += 1
+
+    return groups
+
+
+def group_sparse(int m, int n, int [:] indices, int [:] indptr):
```

Comment:
Potentially might be not, but `np.int32` is the type for sparse matrices `indices` and `indptr`, so it can be changed only on much more fundamental level. 

I think this link is again relevant https://github.com/scikit-learn/scikit-learn/wiki/C-integer-types:-the-missing-manual